dvc/remote/base.py

```diff
@@ -352,7 +363,49 @@ def changed(self, path_info, checksum_info):
         return False
 
     def link(self, from_info, to_info):
-        self.copy(from_info, to_info)
+        self._link(from_info, to_info, self.cache_types)
+
+    def _link(self, from_info, to_info, link_types):
+        assert self.isfile(from_info)
+
+        self.makedirs(to_info.parent)
+
+        self._try_links(from_info, to_info, link_types)
+
+    @slow_link_guard
+    def _try_links(self, from_info, to_info, link_types):
+        i = len(link_types)
+        while i > 0:
+            link_method = getattr(self, link_types[0])
+            try:
+                self._do_link(from_info, to_info, link_method)
+                return
+
+            except DvcException as exc:
+                msg = "Cache type '{}' is not supported: {}"
+                logger.debug(msg.format(link_types[0], str(exc)))
+                del link_types[0]
+                i -= 1
+
+        raise DvcException("no possible cache types left to try out.")
+
+    def _do_link(self, from_info, to_info, link_method):
+        # XXX: We are testing if file exists rather than if file is a link
```

Comment:
Why would we care if that is a link? This check is needed to ensure that we won't be trying to create a link path for which already exists. You can see us using lexists in RemoteLOCAL. But in RemoteSSH, we are using `stat` instead of `lstat` for `exists`, which should be fixed.

Reply:
Agree with `lstat` :+1:

Reply:
The current issue with this one is that `mock-ssh-server` doesn't support `lstat` operation and tests are failing. Need to figure out a way to test this

Reply:
@mroutis It is actually pretty simple to implement, take a look at https://github.com/carletes/mock-ssh-server/blob/master/mockssh/sftp.py#L83 . You can fork it into iterative, modify what is needed and submit a patch. And while the patch is being reviewed, we can use our fork.

Reply:
It's a shame I didn't check this before, @efiop , I ended up using a forked version, indeed :see_no_evil:

Reply:
https://github.com/carletes/mock-ssh-server/pull/9